client/views/view_units.cpp

```diff
@@ -0,0 +1,466 @@
+        }
+      }
+      unit_list_iterate_end;
+    }
+    city_list_iterate_end;
+
+    if (count == 0) {
+      continue;
+    }
+
+    (total_cost) += partial_cost;
+
+    entries[num_entries_used].type = unittype;
+    entries[num_entries_used].count = count;
+    entries[num_entries_used].cost = cost;
+    entries[num_entries_used].total_cost = partial_cost;
+    (num_entries_used)++;
+  }
+  unit_type_iterate_end;
+}
+ */
+
+/**
+   Constructor for units view
+ */
+units_view::units_view() : QWidget()
+{
+  ui.setupUi(this);
+
+  QStringList slist;
+  slist << _("Unit Type") << _("Upgradable ★") << _("In Progress ⚒")
+        << _("Active ⚔") << _("Shield Upkeep") << _("Food Upkeep")
+        << _("Gold Upkeep") << QLatin1String("");
+  ui.units_widget->setColumnCount(slist.count());
+  ui.units_widget->setHorizontalHeaderLabels(slist);
+  ui.units_widget->setSortingEnabled(0);
+  ui.upg_but->setText(_("Upgrade"));
+  ui.upg_but->setDisabled("True");
+  ui.find_but->setText(_("Find Nearest"));
+  ui.find_but->setDisabled("True");
+  ui.disband_but->setText(_("Disband All"));
+  ui.disband_but->setDisabled("True");
+
+  /*connect(ui.upg_but, &QAbstractButton::pressed, this,
+          &units_view::upgrade_units);
+  connect(ui.find_but, &QAbstractButton::pressed, this,
+          &units_view::find_nearest);
+  connect(ui.disband_but, &QAbstractButton::pressed, this,
+          &units_view::disband_units);
+  connect(ui.units_widget->selectionModel(),
+          &QItemSelectionModel::selectionChanged, this,
+          &units_view::selection_changed);*/
+  setLayout(ui.units_layout);
+  queen()->gimmePlace(this, QStringLiteral("UNI"));
+  index = queen()->addGameTab(this);
+}
+
+/**
+   Destructor for units view
+ */
+units_view::~units_view() { queen()->removeRepoDlg(QStringLiteral("UNI")); }
+
+/**
+   Initializes place in tab for units view
+ */
+void units_view::init() { queen()->game_tab_widget->setCurrentIndex(index); }
+
+/**
+   Refresh all widgets for units view
+ */
+void units_view::update_view()
+{
+  struct unit_view_entry unit_entries[U_LAST];
+  int entries_used, i, j, h, total_gold, total_shield, total_food,
+      total_count;
+  QString buf;
+  QTableWidgetItem *item;
+  QTableWidgetItem *item_totals;
+  QFont f = QApplication::font();
+  QFontMetrics fm(f);
+  h = fm.height() + 10;
+
+  ui.units_widget->setRowCount(0);
+  ui.units_widget->clearContents();
+
+  total_count = 0;
+  total_gold = 0;
+  total_shield = 0;
+  total_food = 0;
+  max_row = 0;
+
+  get_units_view_data(unit_entries, &entries_used);
+  for (i = 0; i < entries_used; i++) {
+    struct unit_view_entry *pentry = unit_entries + i;
+    struct unit_type *putype = pentry->type;
+    cid id;
+
+    auto sprite = get_unittype_sprite(tileset, putype, direction8_invalid());
+    id = cid_encode_unit(putype);
+
+    ui.units_widget->insertRow(max_row);
+    for (j = 0; j < 7; j++) {
+      item = new QTableWidgetItem;
+      switch (j) {
```

Comment:
Using a `switch` instead of a loop sounds a bit complicated... The code shared between the iterations is:
```cpp
      item = new QTableWidgetItem;
      ...
      item->setTextAlignment(Qt::AlignVCenter | Qt::AlignHCenter);
      ui.units_widget->setItem(max_row, j, item);
```
I'm not sure it's worth the complexity of a loop and switch...

Reply:
I used code from the economics view that does this loop/switch deal.